modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/XmlProcessor.java

```diff
@@ -0,0 +1,942 @@
+                        xpathExpressions.put(entry.getKey(), str);
+                    } else {
+                        throw new IllegalArgumentException(
+                            "XPath target field ["
+                                + entry.getKey()
+                                + "] must be a string, got ["
+                                + entry.getValue().getClass().getSimpleName()
+                                + "]"
+                        );
+                    }
+                }
+            }
+
+            // Parse namespaces map
+            Map<String, String> namespaces = new HashMap<>();
+            Map<String, Object> namespaceConfig = ConfigurationUtils.readOptionalMap(TYPE, processorTag, config, "namespaces");
+            if (namespaceConfig != null) {
+                for (Map.Entry<String, Object> entry : namespaceConfig.entrySet()) {
+                    if (entry.getValue() instanceof String str) {
+                        namespaces.put(entry.getKey(), str);
+                    } else {
+                        throw new IllegalArgumentException(
+                            "Namespace prefix ["
+                                + entry.getKey()
+                                + "] must have a string URI, got ["
+                                + entry.getValue().getClass().getSimpleName()
+                                + "]"
+                        );
+                    }
+                }
+            }
+
+            // Parse strict_parsing parameter
+            boolean strictParsing = ConfigurationUtils.readBooleanProperty(TYPE, processorTag, config, "strict_parsing", false);
+
+            return new XmlProcessor(
+                processorTag,
+                description,
+                field,
+                targetField,
+                ignoreMissing,
+                toLower,
+                removeEmptyValues,
+                storeXml,
+                removeNamespaces,
+                forceContent,
+                forceArray,
+                xpathExpressions,
+                namespaces,
+                strictParsing
+            );
+        }
+    }
+
+    /**
+     * Main XML parsing method that converts XML to JSON and optionally extracts XPath values.
+     * Uses streaming SAX parser with optional DOM building for XPath processing.
+     *
+     * @param document the ingest document to modify with parsed results
+     * @param xmlString the XML string to parse (should be trimmed)
+     * @throws Exception if XML parsing fails
+     */
+    private void parseXmlAndXPath(IngestDocument document, String xmlString) throws Exception {
+        if (Strings.hasText(xmlString) == false) {
+            return;
+        }
+
+        // Determine if we need DOM for XPath processing
+        boolean needsDom = xpathExpressions.isEmpty() == false;
+
+        // Use the appropriate pre-configured SAX parser factory
+        SAXParserFactory factory = selectSaxParserFactory();
```

Comment:
I don't think this needs to be done per-document either. My casual read of it is that the parser factory could be a final variable selected as configuration parse time (that is, at processor construction time).

For this and my previous comment, it's less about **performance** than it is about making the code obvious.

Things that do not **have** to be executed on a **per-document** basis shouldn't be.